experimental/StandardFiniteFields/StandardFiniteFields.jl

```diff
@@ -0,0 +1,462 @@
+
+
+# _attributes = [
+    # :is_standard_finite_field      - bool
+    # :is_standard_prime_field       - bool
+    # :primitive_powers_in_tower_basis - Matrix{FinFieldElem}
+    # :tower_basis                  - Matrix{FFieldElem} = inv(:primitive_powers_in_tower_basis)
+    # NOTE I guess we used ZZ ring elements for these instead of Int; does it matter?
+    # :steinitz_prime_degree         - Dict{Int, Dict{Int, Int}}
+    # :standard_extensions          - Dict{Int, FinField} ]
+function set_standard_prime_field!(F::PrimeField)
+    get_attribute!(F, :is_standard_prime_field) do
+        set_attribute!(F, :standard_extensions, Dict{ZZRingElem, FinField}(1 => F))
+        set_attribute!(F, :primitive_powers_in_tower_basis, identity_matrix(F, 1))
+        true
+    end
+end
+function set_standard_finite_field!(F::FinField)
+    set_attribute!(F, :is_standard_finite_field, true)
+end
+function set_primitive_powers_in_tower_basis!(F::FinField, M::PrimeFieldMatrix)
+    set_attribute!(F, :primitive_powers_in_tower_basis, M)
+    set_attribute!(F, :tower_basis, inv(M))
+end
+function set_steinitz_prime_degree!(F::FinField, r::IntegerUnion, k::IntegerUnion, nr::IntegerUnion)
+    spd = get_attribute!(F, :steinitz_prime_degree, Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}())
+    spdr = get!(spd, r, Dict{ZZRingElem, ZZRingElem})
+    spdr[k] = nr
+end
+function set_steinitz_prime_degree!(f::Function, F::FinField, r::IntegerUnion, k::IntegerUnion)
+    spd = get_attribute!(F, :steinitz_prime_degree, Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}())
+    spdr = get!(spd, r, Dict{ZZRingElem, ZZRingElem})
+    spd[r][k] = f()
+end
+function set_standard_extension!(F::PrimeField, n::IntegerUnion, K::FinField)
+    ext = get_attribute!(F, :standard_extensions, Dict{ZZRingElem, FinField}())
+    ext[n] = K
+end
+
+
+function is_standard_prime_field(F::PrimeField)
+    get_attribute(F, :is_standard_prime_field, false)
+end
+function is_standard_finite_field(F::FinField)
+    get_attribute(F, :is_standard_finite_field, false)
+end
+function primitive_powers_in_tower_basis(F::FinField)
+    get_attribute(F, :primitive_powers_in_tower_basis, nothing)
+end
+function tower_basis(F::FinField)
+    get_attribute(F, :tower_basis, nothing)
+end
+function get_steinitz_prime_degree(F::FinField, r::IntegerUnion, k::IntegerUnion)
+    spd = get_attribute(F, :steinitz_prime_degree, nothing)
+    spdr = get(spd, r, nothing)
+    get(spdr, k, nothing)
+end
+function get_steinitz_prime_degree!(f::Function, F::FinField, r::IntegerUnion, k::IntegerUnion)
+    spd = get_attribute!(F, :steinitz_prime_degree, Dict{ZZRingElem, Dict{ZZRingElem, ZZRingElem}}() )
+    spdr = get!(spd, r, Dict{ZZRingElem, ZZRingElem}())
+    get!(spdr, k, f())
+end
+function get_standard_extensions(F::PrimeField)
+  get_attribute(F, :standard_extensions, nothing)
+end
+function get_standard_extensions!(F::PrimeField)
+  get_attribute(F, :standard_extensions, Dict{ZZRingElem, FinField}())
```

Comment:
Note that this always allocate a `Dict{ZZRingElem, FinField}`, even if you don't need it because the `:standard_extensions` already exists. Hence it is better to do this:
```suggestion
  get_attribute(F, :standard_extensions) do
    Dict{ZZRingElem, FinField}()
  end
```
or even this
```suggestion
  get_attribute(Dict{ZZRingElem, FinField}, F, :standard_extensions)
```